src/librustc/ty/mod.rs

```diff
@@ -771,25 +784,53 @@ pub struct Generics {
 }
 
 impl<'a, 'gcx, 'tcx> Generics {
-    pub fn parent_count(&self) -> usize {
-        self.parent_regions as usize + self.parent_types as usize
+    pub fn count(&self) -> usize {
+        self.parent_count + self.params.len()
+    }
+
+    pub fn lifetimes(&self) -> impl DoubleEndedIterator<Item = &RegionParameterDef> {
+        self.params.iter().filter_map(|p| {
+            if let GenericParam::Lifetime(lt) = p {
+                Some(lt)
+            } else {
+                None
+            }
+        })
     }
 
-    pub fn own_count(&self) -> usize {
-        self.regions.len() + self.types.len()
+    pub fn types(&self) -> impl DoubleEndedIterator<Item = &TypeParameterDef> {
+        self.params.iter().filter_map(|p| {
+            if let GenericParam::Type(ty) = p {
+                Some(ty)
+            } else {
+                None
+            }
+        })
     }
 
-    pub fn count(&self) -> usize {
-        self.parent_count() + self.own_count()
+    pub fn has_type_parameters(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
```

Comment:
This appears to be used in monomorphization decisions, but we might want something different (e.g. https://github.com/rust-lang/rust/pull/48139#issuecomment-373985951), and this would have to also take into account lifetime parameters - so how about naming this `requires_monomorphization` and looping over params manually?